packages/muelu/src/Misc/MueLu_InterfaceAggregationFactory_def.hpp

```diff
@@ -0,0 +1,191 @@
+  validParamList->set<RCP<std::map<LocalOrdinal, LocalOrdinal>>>("DualNodeID2PrimalNodeID", null, "");
+
+  return validParamList;
+}
+
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void InterfaceAggregationFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::DeclareInput(Level &currentLevel) const
+{
+  Input(currentLevel, "A");
+  Input(currentLevel, "Aggregates");
+
+  if (currentLevel.GetLevelID() != 0)
+    currentLevel.DeclareInput("DualNodeID2PrimalNodeID", NoFactory::get());
+}
+
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void InterfaceAggregationFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Build(Level &currentLevel) const
+{
+  typedef Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node> Map;
+  typedef Aggregates<LocalOrdinal, GlobalOrdinal, Node> Aggregates;
+  typedef std::map<LocalOrdinal, LocalOrdinal> Dual2Primal_type;
+
+  const char prefix[] = "MueLu::InterfaceAggregationFactory::Build: ";
+  const ParameterList &pL = GetParameterList();
+
+  FactoryMonitor m(*this, "Build", currentLevel);
+
+  RCP<Matrix> A = Get<RCP<Matrix>>(currentLevel, "A");
+  RCP<Aggregates> aggs00 = Get<RCP<Aggregates>>(currentLevel, "Aggregates");
+  ArrayRCP<LocalOrdinal> vertex2AggIdInput = aggs00->GetVertex2AggId()->getDataNonConst(0);
+
+  ArrayRCP<LocalOrdinal> procWinnerInput = aggs00->GetProcWinner()->getDataNonConst(0);
+
+  RCP<Dual2Primal_type> lagr2dof;
+  if (currentLevel.GetLevelID() == 0)
+  {
+    lagr2dof = pL.get<RCP<Dual2Primal_type>>("DualNodeID2PrimalNodeID");
+
+    TEUCHOS_TEST_FOR_EXCEPTION(lagr2dof.is_null(), std::runtime_error, prefix << " MueLu requires a DualNodeID2PrimalNodeID map.");
+  }
+  else
+    lagr2dof = currentLevel.Get<RCP<Dual2Primal_type>>("DualNodeID2PrimalNodeID", NoFactory::get());
+
+  const LocalOrdinal nDOFPerDualNode = pL.get<LocalOrdinal>("number of DOFs per dual node");
+
+  RCP<const Map> aggRangeMap = A->getRangeMap();
+  const size_t myRank = aggRangeMap->getComm()->getRank();
+
+  LocalOrdinal globalNumDualNodes = aggRangeMap->getGlobalNumElements() / nDOFPerDualNode;
+  LocalOrdinal numDualNodes = aggRangeMap->getNodeNumElements() / nDOFPerDualNode;
+
+  TEUCHOS_TEST_FOR_EXCEPTION(numDualNodes != LocalOrdinal(lagr2dof->size()), std::runtime_error, prefix << " MueLu requires the range map and the DualNodeID2PrimalNodeID map to be compatible.");
+
+  RCP<const Map> aggVertexMap;
+
+  if (nDOFPerDualNode == 1)
+    aggVertexMap = aggRangeMap;
+  else
+  {
+    GlobalOrdinal indexBase = aggRangeMap->getIndexBase();
+    auto comm = aggRangeMap->getComm();
+    std::vector<GlobalOrdinal> myDualNodes = {};
+
+    for (size_t i = 0; i < aggRangeMap->getNodeNumElements(); i += nDOFPerDualNode)
+      myDualNodes.push_back((aggRangeMap->getGlobalElement(i) - indexBase) / nDOFPerDualNode + indexBase);
+
+    aggVertexMap = rcp(new const Xpetra::TpetraMap<LocalOrdinal, GlobalOrdinal, Node>(globalNumDualNodes, myDualNodes, indexBase, comm));
```

Comment:
At the beginning of this function, you typedef `Xpetra::Map` (which can be an Epetra or Tpetra map if I'm not mistaken), but here you explicitly create an `Xpetra::TpetraMap`. Is there a specific reason for that? What about running this code with Epetra?

Reply:
You are right! Thanks!